File: NeptuneCSample/NeptuneCSample_VBNET/PopupTriggerForm.vb
Imports NeptuneCSample_VBNET.NeptuneC_Interface

Public Class PopupTriggerForm
    Private m_refMainForm As NeptuneCSampleForm = Nothing

    Public Sub New(ByVal refMainForm As NeptuneCSampleForm)
        InitializeComponent()
        Me.m_refMainForm = refMainForm
    End Sub

    Private Sub InitControls()

    End Sub

    Public Sub UpdateForm()
        m_ckTriggerEnable.Checked = False
        m_cbTriggerMode.Items.Clear()
        m_cbTriggerSource.Items.Clear()
        m_cbTriggerPolarity.Items.Clear()
        m_edTriggerParam.Minimum = 0
        m_edTriggerParam.Maximum = 0
        m_edTriggerParam.Value = 0
        m_btnSWTrigger.Enabled = False
        m_stcParamRange.Text = "(Min ~ Max)"
        If (Me.m_refMainForm.m_pCameraHandle <> IntPtr.Zero) Then
            Dim stTriggerInfo As NEPTUNE_TRIGGER_INFO = New NEPTUNE_TRIGGER_INFO
            Dim emErr As ENeptuneError = NeptuneC.ntcGetTriggerInfo(Me.m_refMainForm.m_pCameraHandle, stTriggerInfo)
            If (emErr = ENeptuneError.NEPTUNE_ERR_Success) Then
                If (stTriggerInfo.bSupport = ENeptuneBoolean.NEPTUNE_BOOL_TRUE) Then
                    Dim nFlag As Integer = CType(stTriggerInfo.nModeFlag, Integer)
                    Dim i As Integer = 0
                    For i = 0 To 30
                        If (nFlag And &H1) = &H1 Then
                            m_cbTriggerMode.Items.Add(New ItemData(("Mode" + i.ToString()), i))
                        End If

                        nFlag = (nFlag >> 1)
                    Next

                    nFlag = CType((stTriggerInfo.nSourceFlag >> CType(ENeptuneTriggerSource.NEPTUNE_TRIGGER_SOURCE_LINE1, Integer)), Integer)
                    If ((nFlag And 1) _
                                = 1) Then
                        m_cbTriggerSource.Items.Add(New ItemData("Line1", CType(ENeptuneTriggerSource.NEPTUNE_TRIGGER_SOURCE_LINE1, Integer)))
                    End If

                    nFlag = CType((stTriggerInfo.nSourceFlag >> CType(ENeptuneTriggerSource.NEPTUNE_TRIGGER_SOURCE_SW, Integer)), Integer)
                    If ((nFlag And 1) _
                                = 1) Then
                        m_cbTriggerSource.Items.Add(New ItemData("Software", CType(ENeptuneTriggerSource.NEPTUNE_TRIGGER_SOURCE_SW, Integer)))
                    End If

                    nFlag = CType((stTriggerInfo.nPolarityFlag >> CType(ENeptunePolarity.NEPTUNE_POLARITY_RISINGEDGE, Integer)), Integer)
                    If ((nFlag And 1) _
                                = 1) Then
                        m_cbTriggerPolarity.Items.Add(New ItemData("Rising Edge", CType(ENeptunePolarity.NEPTUNE_POLARITY_RISINGEDGE, Integer)))
                    End If

                    nFlag = CType((stTriggerInfo.nPolarityFlag >> CType(ENeptunePolarity.NEPTUNE_POLARITY_FALLINGEDGE, Integer)), Integer)
                    If ((nFlag And 1) _
                                = 1) Then
                        m_cbTriggerPolarity.Items.Add(New ItemData("Falling Edge", CType(ENeptunePolarity.NEPTUNE_POLARITY_FALLINGEDGE, Integer)))
                    End If

                    nFlag = CType((stTriggerInfo.nPolarityFlag >> CType(ENeptunePolarity.NEPTUNE_POLARITY_ANYEDGE, Integer)), Integer)
                    If ((nFlag And 1) _
                                = 1) Then
                        m_cbTriggerPolarity.Items.Add(New ItemData("Any Edge", CType(ENeptunePolarity.NEPTUNE_POLARITY_ANYEDGE, Integer)))
                    End If

                    nFlag = CType((stTriggerInfo.nPolarityFlag >> CType(ENeptunePolarity.NEPTUNE_POLARITY_LEVELHIGH, Integer)), Integer)
                    If ((nFlag And 1) _
                                = 1) Then
                        m_cbTriggerPolarity.Items.Add(New ItemData("High Level", CType(ENeptunePolarity.NEPTUNE_POLARITY_LEVELHIGH, Integer)))
                    End If

                    nFlag = CType((stTriggerInfo.nPolarityFlag >> CType(ENeptunePolarity.NEPTUNE_POLARITY_LEVELLOW, Integer)), Integer)
                    If ((nFlag And 1) _
                                = 1) Then
                        m_cbTriggerPolarity.Items.Add(New ItemData("Low Level", CType(ENeptunePolarity.NEPTUNE_POLARITY_LEVELLOW, Integer)))
                    End If

                    m_edTriggerParam.Minimum = stTriggerInfo.nParamMin
                    m_edTriggerParam.Maximum = stTriggerInfo.nParamMax
                    m_stcParamRange.Text = ("(" _
                                + (m_edTriggerParam.Minimum.ToString() + (" ~ " _
                                + (m_edTriggerParam.Maximum.ToString() + ")"))))
                    Dim stTrigger As NEPTUNE_TRIGGER = New NEPTUNE_TRIGGER
                    emErr = NeptuneC.ntcGetTrigger(Me.m_refMainForm.m_pCameraHandle, stTrigger)
                    If (emErr = ENeptuneError.NEPTUNE_ERR_Success) Then
                        If stTrigger.OnOff = ENeptuneBoolean.NEPTUNE_BOOL_TRUE Then
                            m_ckTriggerEnable.Checked = True
                        Else
                            m_ckTriggerEnable.Checked = False
                        End If

                        For i = 0 To m_cbTriggerMode.Items.Count - 1
                            If (CType(m_cbTriggerMode.Items(i), ItemData).m_nValue = CType(stTrigger.Mode, Integer)) Then
                                m_cbTriggerMode.SelectedIndex = i
                                Exit For
                            End If
                        Next

                        For i = 0 To m_cbTriggerSource.Items.Count - 1
                            If (CType(m_cbTriggerSource.Items(i), ItemData).m_nValue = CType(stTrigger.Source, Integer)) Then
                                m_cbTriggerSource.SelectedIndex = i
                                Exit For
                            End If
                        Next

                        For i = 0 To m_cbTriggerPolarity.Items.Count - 1
                            If (CType(m_cbTriggerPolarity.Items(i), ItemData).m_nValue = CType(stTrigger.Polarity, Integer)) Then
                                m_cbTriggerPolarity.SelectedIndex = i
                                Exit For
                            End If
                        Next

                        m_edTriggerParam.Value = stTrigger.nParam
                        m_btnSWTrigger.Enabled = ((stTrigger.Source = ENeptuneTriggerSource.NEPTUNE_TRIGGER_SOURCE_SW) _
                                    AndAlso (stTrigger.OnOff = ENeptuneBoolean.NEPTUNE_BOOL_TRUE))
                    Else
                        Me.m_refMainForm.InsertLog(EM_LOG_LEVEL.TYPE_ERROR, "ntcGetTrigger Failed.", emErr)
                    End If

                End If

            Else
                Me.m_refMainForm.InsertLog(EM_LOG_LEVEL.TYPE_ERROR, "ntcGetTriggerInfo Failed.", emErr)
            End If

        End If

    End Sub

    Private Sub PopupTriggerForm_Load(sender As System.Object, e As System.EventArgs) Handles MyBase.Load
        Me.FormBorderStyle = FormBorderStyle.FixedSingle
        Me.InitControls()
        Me.UpdateForm()
    End Sub

    Private Sub PopupTriggerForm_FormClosed(sender As System.Object, e As System.Windows.Forms.FormClosedEventArgs) Handles MyBase.FormClosed
        Me.m_refMainForm.m_formPopupTrigger = Nothing
    End Sub

    Private Sub m_btnApply_Click(sender As System.Object, e As System.EventArgs) Handles m_btnApply.Click
        If (Me.m_refMainForm.m_pCameraHandle <> IntPtr.Zero) Then
            Dim stTrigger As NEPTUNE_TRIGGER = New NEPTUNE_TRIGGER
            If m_ckTriggerEnable.Checked = True Then
                stTrigger.OnOff = ENeptuneBoolean.NEPTUNE_BOOL_TRUE
            Else
                stTrigger.OnOff = ENeptuneBoolean.NEPTUNE_BOOL_FALSE
            End If

            stTrigger.Mode = CType(CType(m_cbTriggerMode.SelectedItem, ItemData).m_nValue, ENeptuneTriggerMode)
            stTrigger.Source = CType(CType(m_cbTriggerSource.SelectedItem, ItemData).m_nValue, ENeptuneTriggerSource)
            stTrigger.Polarity = CType(CType(m_cbTriggerPolarity.SelectedItem, ItemData).m_nValue, ENeptunePolarity)
            stTrigger.nParam = CType(m_edTriggerParam.Value, System.UInt16)
            Dim emErr As ENeptuneError = NeptuneC.ntcSetTrigger(Me.m_refMainForm.m_pCameraHandle, stTrigger)
            If (emErr = ENeptuneError.NEPTUNE_ERR_Success) Then
                m_btnSWTrigger.Enabled = ((stTrigger.Source = ENeptuneTriggerSource.NEPTUNE_TRIGGER_SOURCE_SW) _
                            AndAlso (stTrigger.OnOff = ENeptuneBoolean.NEPTUNE_BOOL_TRUE))
            Else
                Me.UpdateForm()
                Me.m_refMainForm.InsertLog(EM_LOG_LEVEL.TYPE_ERROR, "ntcSetTrigger Failed.", emErr)
            End If

        End If
    End Sub

    Private Sub m_btnSWTrigger_Click(sender As System.Object, e As System.EventArgs) Handles m_btnSWTrigger.Click
        If (Me.m_refMainForm.m_pCameraHandle <> IntPtr.Zero) Then
            NeptuneC.ntcRunSWTrigger(Me.m_refMainForm.m_pCameraHandle)
        End If
    End Sub
End Class

File: NeptuneCSample/NeptuneCSample_CSharp/PopupTriggerForm.cs
using System;
using System.Collections.Generic;
using System.ComponentModel;
using System.Data;
using System.Drawing;
using System.Linq;
using System.Text;
using System.Windows.Forms;

using NeptuneC_Interface;

namespace NeptuneCSample_CSharp
{
    public partial class PopupTriggerForm : Form
    {
        private NeptuneCSampleForm m_refMainForm = null;

        public PopupTriggerForm(NeptuneCSampleForm refMainForm)
        {
            InitializeComponent();

            m_refMainForm = refMainForm;
        }

        private void InitControls()
        {
        }

        public void UpdateForm()
        {
            m_ckTriggerEnable.Checked = false;
            m_cbTriggerMode.Items.Clear();
            m_cbTriggerSource.Items.Clear();
            m_cbTriggerPolarity.Items.Clear();
            m_edTriggerParam.Minimum = 0;
            m_edTriggerParam.Maximum = 0;
            m_edTriggerParam.Value = 0;
            m_btnSWTrigger.Enabled = false;
            m_stcParamRange.Text = "(Min ~ Max)";

            if (m_refMainForm.m_pCameraHandle != IntPtr.Zero)
            {
                NEPTUNE_TRIGGER_INFO stTriggerInfo = new NEPTUNE_TRIGGER_INFO();
                ENeptuneError emErr = NeptuneC.ntcGetTriggerInfo(m_refMainForm.m_pCameraHandle, ref stTriggerInfo);
                if (emErr == ENeptuneError.NEPTUNE_ERR_Success)
                {
                    if (stTriggerInfo.bSupport == ENeptuneBoolean.NEPTUNE_BOOL_TRUE)
                    {
                        int nFlag = (int)stTriggerInfo.nModeFlag;
                        for (int i = 0; i < 31; i++)
                        {
                            if ((nFlag & 0x01) == 0x01)
                            {
                                m_cbTriggerMode.Items.Add(new ItemData("Mode" + i, i));
                            }
                            nFlag = (nFlag >> 1);
                        }

                        nFlag = (int)(stTriggerInfo.nSourceFlag >> (int)ENeptuneTriggerSource.NEPTUNE_TRIGGER_SOURCE_LINE1);
                        if ((nFlag & 0x01) == 0x01)
                        {
                            m_cbTriggerSource.Items.Add(new ItemData("Line1", (int)ENeptuneTriggerSource.NEPTUNE_TRIGGER_SOURCE_LINE1));
                        }
                        nFlag = (int)(stTriggerInfo.nSourceFlag >> (int)ENeptuneTriggerSource.NEPTUNE_TRIGGER_SOURCE_SW);
                        if ((nFlag & 0x01) == 0x01)
                        {
                            m_cbTriggerSource.Items.Add(new ItemData("Software", (int)ENeptuneTriggerSource.NEPTUNE_TRIGGER_SOURCE_SW));
                        }
                        
                        nFlag = (int)(stTriggerInfo.nPolarityFlag >> (int)ENeptunePolarity.NEPTUNE_POLARITY_RISINGEDGE);
                        if ((nFlag & 0x01) == 0x01)
                        {
                            m_cbTriggerPolarity.Items.Add(new ItemData("Rising Edge", (int)ENeptunePolarity.NEPTUNE_POLARITY_RISINGEDGE));
                        }
                        nFlag = (int)(stTriggerInfo.nPolarityFlag >> (int)ENeptunePolarity.NEPTUNE_POLARITY_FALLINGEDGE);
                        if ((nFlag & 0x01) == 0x01)
                        {
                            m_cbTriggerPolarity.Items.Add(new ItemData("Falling Edge", (int)ENeptunePolarity.NEPTUNE_POLARITY_FALLINGEDGE));
                        }
                        nFlag = (int)(stTriggerInfo.nPolarityFlag >> (int)ENeptunePolarity.NEPTUNE_POLARITY_ANYEDGE);
                        if ((nFlag & 0x01) == 0x01)
                        {
                            m_cbTriggerPolarity.Items.Add(new ItemData("Any Edge", (int)ENeptunePolarity.NEPTUNE_POLARITY_ANYEDGE));
                        }
                        nFlag = (int)(stTriggerInfo.nPolarityFlag >> (int)ENeptunePolarity.NEPTUNE_POLARITY_LEVELHIGH);
                        if ((nFlag & 0x01) == 0x01)
                        {
                            m_cbTriggerPolarity.Items.Add(new ItemData("High Level", (int)ENeptunePolarity.NEPTUNE_POLARITY_LEVELHIGH));
                        }
                        nFlag = (int)(stTriggerInfo.nPolarityFlag >> (int)ENeptunePolarity.NEPTUNE_POLARITY_LEVELLOW);
                        if ((nFlag & 0x01) == 0x01)
                        {
                            m_cbTriggerPolarity.Items.Add(new ItemData("Low Level", (int)ENeptunePolarity.NEPTUNE_POLARITY_LEVELLOW));
                        }

                        m_edTriggerParam.Minimum = stTriggerInfo.nParamMin;
                        m_edTriggerParam.Maximum = stTriggerInfo.nParamMax;
                        m_stcParamRange.Text = "(" +  m_edTriggerParam.Minimum + " ~ " + m_edTriggerParam.Maximum + ")";

                        NEPTUNE_TRIGGER stTrigger = new NEPTUNE_TRIGGER();
                        emErr = NeptuneC.ntcGetTrigger(m_refMainForm.m_pCameraHandle, ref stTrigger);
                        if (emErr == ENeptuneError.NEPTUNE_ERR_Success)
                        {
                            m_ckTriggerEnable.Checked = (stTrigger.OnOff == ENeptuneBoolean.NEPTUNE_BOOL_TRUE ? true : false);
                            
                            for (int i = 0; i < m_cbTriggerMode.Items.Count; i++)
                            {
                                if (((ItemData)m_cbTriggerMode.Items[i]).m_nValue == (int)stTrigger.Mode)
                                {
                                    m_cbTriggerMode.SelectedIndex = i;
                                    break;
                                }
                            }

                            for (int i = 0; i < m_cbTriggerSource.Items.Count; i++)
                            {
                                if (((ItemData)m_cbTriggerSource.Items[i]).m_nValue == (int)stTrigger.Source)
                                {
                                    m_cbTriggerSource.SelectedIndex = i;
                                    break;
                                }
                            }

                            for (int i = 0; i < m_cbTriggerPolarity.Items.Count; i++)
                            {
                                if (((ItemData)m_cbTriggerPolarity.Items[i]).m_nValue == (int)stTrigger.Polarity)
                                {
                                    m_cbTriggerPolarity.SelectedIndex = i;
                                    break;
                                }
                            }

                            m_edTriggerParam.Value = stTrigger.nParam;

                            m_btnSWTrigger.Enabled = ((stTrigger.Source == ENeptuneTriggerSource.NEPTUNE_TRIGGER_SOURCE_SW) && (stTrigger.OnOff == ENeptuneBoolean.NEPTUNE_BOOL_TRUE));
                        }
                        else
                        {
                            m_refMainForm.InsertLog(EM_LOG_LEVEL.TYPE_ERROR, "ntcGetTrigger Failed.", emErr);
                        }
                    }
                }
                else
                {
                    m_refMainForm.InsertLog(EM_LOG_LEVEL.TYPE_ERROR, "ntcGetTriggerInfo Failed.", emErr);
                }
            }
        }

        private void PopupTriggerForm_Load(object sender, EventArgs e)
        {
            this.FormBorderStyle = FormBorderStyle.FixedSingle;

            InitControls();
            UpdateForm();
        }

        private void PopupTriggerForm_FormClosed(object sender, FormClosedEventArgs e)
        {
            m_refMainForm.m_formPopupTrigger = null;
        }

        private void m_btnApply_Click(object sender, EventArgs e)
        {
            if (m_refMainForm.m_pCameraHandle != IntPtr.Zero)
            {
                NEPTUNE_TRIGGER stTrigger = new NEPTUNE_TRIGGER();
                stTrigger.OnOff = (m_ckTriggerEnable.Checked ? ENeptuneBoolean.NEPTUNE_BOOL_TRUE : ENeptuneBoolean.NEPTUNE_BOOL_FALSE);
                stTrigger.Mode = (ENeptuneTriggerMode)((ItemData)m_cbTriggerMode.SelectedItem).m_nValue;
                stTrigger.Source = (ENeptuneTriggerSource)((ItemData)m_cbTriggerSource.SelectedItem).m_nValue;
                stTrigger.Polarity = (ENeptunePolarity)((ItemData)m_cbTriggerPolarity.SelectedItem).m_nValue;
                stTrigger.nParam = (ushort)m_edTriggerParam.Value;
                ENeptuneError emErr = NeptuneC.ntcSetTrigger(m_refMainForm.m_pCameraHandle, stTrigger);
                if (emErr == ENeptuneError.NEPTUNE_ERR_Success)
                {
                    m_btnSWTrigger.Enabled = ((stTrigger.Source == ENeptuneTriggerSource.NEPTUNE_TRIGGER_SOURCE_SW) && (stTrigger.OnOff == ENeptuneBoolean.NEPTUNE_BOOL_TRUE));
                }
                else
                {
                    UpdateForm();
                    m_refMainForm.InsertLog(EM_LOG_LEVEL.TYPE_ERROR, "ntcSetTrigger Failed.", emErr);
                }
            }
        }

        private void m_btnSWTrigger_Click(object sender, EventArgs e)
        {
            if (m_refMainForm.m_pCameraHandle != IntPtr.Zero)
            {
                NeptuneC.ntcRunSWTrigger(m_refMainForm.m_pCameraHandle);
            }
        }
    }
}
